### ISSUE TITLE
Split pointer enter, exit, and hover callbacks from Listener

### ISSUE THREAD

**Issue by dkwingsmt** [flutter/flutter#36085]
## TL;DR
It is proposed that pointer enter, exit, and hover callbacks are split from `Listener` into a new class `MouseListener`. It is mainly because mouse tracking is unable to support certain features provided by `Listener`, while also because when `Listener` is explicitly used, it is almost always only for mouse events.

### Unable to support `HitTestBehavior.deferToChild`
`Listener` is designed to be a all-in-one class for pointer-related callbacks. Technically it consists of two categories of callbacks: Mouse tracking, including by `onPointer{Enter,Exit,Hover}` events, are implemented with layer-based hit testing, while pointer events listening, `onPointer{Down,Up,Move}`, are implemented with render-box-based hit testing.

One of the options `Listener` provide is `HitTestBehavior behavior`, which controls the behavior of hit testing. However, since this option is directly passed to `RenderProxyBoxWithHitTestBehavior`, it does not affect layers. Having an option that only affects part of the callbacks is definitely confusing and undesired.

We can try to simulate the behavior in layers (which is planned), but it is technically impossible to simulate `deferToChild`, since child information is related with render objects and is lost when converted to layers. 

By splitting `onPointer{Enter,Exit,Hover}` into a separate class, we can provide a different type of behavior that has only two options to avoid this confusion.

### It is often used alone
`Listener.onPointer{Down,Up,Move}` are rarely used directly, instead developers are supposed to use `GestureDetector` when possible, or to create a custom gesture recognizer with `RawGestureDetector`.

On the other hand, `Listener.onPointer{Enter,Exit,Hover}` are currently the only way to use mouse tracking events. This leads to a common code pattern as the following snippet, where `Listener` is only used for `onPointer{Enter,Exit,Hover}`:
```dart
  @override
  Widget build(BuildContext context) {
    return Listener(
      onPointerEnter: _handlePointerEnter,
      onPointerExit: _handlePointerExit,
      child: GestureDetector(
        onTapDown: _handleTapDown,
        onTap: _handleTap,
        ...
        child: widget.child,
      ),
    );
  }
```

Although we squashed 2 separate subsystems into `Listener`, when it's used by developers, only one of them is utilized. Therefore it might be a better idea to split them for modularity.

## Proposal
- Split the mouse tracking part from `Listener` and `RenderPointerListener` into new classes, tentatively named `MouseListener` and `RenderMouseListener`
- `Listener.onPointer{Enter,Exit,Hover}` are preserved for backward compatibility and marked as deprecated.

**Comment by dkwingsmt** [comment]
+cc @gspencergoog @Hixie

**Comment by gspencergoog** [comment]
This seems like a good proposal to me. For people who do check for pointer exit/enter, it is a big breaking change, but it would make it easier to understand.

**Comment by Hixie** [comment]
I would call this something like MouseRegion and also add the mouse cursor control to this class.
Other than that I agree in principle with this offering.
I worry that this is a hard break. We should investigate if there's a way to add this feature while still leaving the old API as @deprecated until the next stable release. (cc @amirh)

**Comment by mdebbar** [comment]
The proposal looks good, thanks for writing it! My only comment would be that the name `Mouse` is too generic and too specific at the same time. It's too generic in that it doesn't say what it does with the "mouse". And it's too specific in that it's tied to a single device kind.

From my experience looking at Flutter code, I haven't seen any APIs that say "Touch" or "Stylus" for example. To match the rest of Flutter's naming, we might want to consider using a name that better describes the intention of this class/widget. Maybe something like `HoverArea` or `Hoverable`?

**Comment by dkwingsmt** [comment]
@mdebbar Thanks for the suggestion! The name is indeed a concern, and I'd love to discuss it.

I was originally considering Hixie's suggestion `MouseRegion` but then switched to the more generic name `Mouse` (which is of course not the final decision). 
I chose a generic name when writing this proposal because Flutter has had several similarly generic widget names, e.g. `Semantic`, `Listener`, and `Focus`. While `MouseRegion` or `MouseArea` might be better than `Mouse`, I really like `HoverArea` or `Hoverable` as they're great names for a widget. I'll probably switch to them if no better names are brought up.

Regarding being too specific, Flutter has had a few widgets specifying "mouse", such as `MouseTracker`  or `type: PointerDeviceKind.mouse`. I think it shows that Flutter is already pretty committed to it (but I agree that it would be better to avoid the problem anyway)

**Comment by Hixie** [comment]
The MouseRegion proposal was based on the idea that what we're doing here is specific to mouse regions (whether hovering or dragging) and not hovering (e.g. it doesn't apply to pointers that hover, it applies to mice even when they're dragging, not hovering). MouseArea seems ok too, though we tend to use the word "region" rather than "area" in talking about Flutter's API generally.

**Comment by chunhtai** [comment]
This is not related to this change, but can a stylus do hover? I remember some devices can detect stylus that are really close to screen but not actually touching it.

**Comment by dkwingsmt** [comment]
@chunhtai I'm aware. I think the idea right now is to commit to mice due to the following reasons:
1. If I remember correctly Flutter is not fully ready to support hoverable styluses yet, therefore it's premature to consider it right now. For example, should hoverable styluses considered a separate device? What if the OS just simulates mouse events out of them?
2. There are some other places that have committed to mice, and switching to also supporting styluses already requires breaking changes.

Since this change is currently blocking a few other changes, such as layer-based hit test behavior and mouse cursor, I think it's a fair choice to commit to mice for simplicity.

**Comment by tvolkert** [comment]
The `Listener` hooks have been deprecated for three stable releases now.  It should be safe to remove them now.

https://github.com/flutter/flutter/blob/d1583fc39b8044d8e92fed63a01e0b523ce248b7/packages/flutter/lib/src/widgets/basic.dart#L5788-L5804
